summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala

```diff
@@ -107,13 +132,13 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
     if (pendingFuturesCount > maxWaitingFuturesCount) {
       // Too many futures waiting, let's clear.
       outstandingFutures.dequeueAll(_.isDefined)
-      val toForce = outstandingFutures.trimTo(maxWaitingFuturesCount).toIndexedSeq
-      if (toForce.nonEmpty) {
+      val toClear = outstandingFutures.size - maxWaitingFuturesCount
+      if (toClear > 0) {
         try {
-          Await.ready(Future.collect(toForce), maxWaitingTime.get)
+          Await.ready(AsyncBase.waitN(outstandingFutures, toClear), maxWaitingTime.get)
```

Comment:
I'm not sure dequeueing is much more expensive than iterating. Iterating over a mutable collection is problematic. I think dequeue + requeue is clearer semantic.

With your previous check that we only do this when the number of finished futures is high, I don't think it will be expensive. I think we should try and see if perf is not good (maybe use yourkit) rather than possibly make it more confusing iterating through a mutable collection.